lib/src/types.ts

```diff
@@ -431,6 +456,28 @@ export function possibleRootTypes(
       []
     );
   }
+
+  if (isIntersectionType(type)) {
+    const concreteTypes = type.types.reduce<ConcreteType[]>((acc, curr) => {
+      return acc.concat(possibleRootTypes(curr, typeTable));
+    }, []);
+    // if they are all objects then reduce them into one object
+    // to allow for proper serialization
+    if (concreteTypes.every(isObjectType)) {
+      return [
+        {
+          kind: TypeKind.OBJECT,
+          properties: concreteTypes.reduce<ObjectPropertiesType[]>(
+            (acc, curr) => {
+              return acc.concat(...curr.properties);
```

Comment:
Noticed this might have a similar problem to `doesInterfaceEvaluatesToNever` for cases that have overlapping or duplicate properties:

```ts
type A = { x: string, y: string } & { x: string, z: string };
type B = { x: string, y: string } & { x: "abc", z: string };
```

In this case we should resolve to the narrowest type:
```ts
type A = { x: string, y: string, z: string };
type B = { x: "abc", y: string, z: string };
```

Reply:
Looks like providing a narrow type might require more changes in the `type-parser.ts` to correctly drill down the type, which might get more complicated for some of the reference types 🤔

Reply:
hey @lfportal I am able to find the narrow type and drill down to it, however when it starts being a reference chain, it starts to get a little complicated and starts to include elements which are not meant to be there. Specifically see the following example:

```
type ABC = M
type M = P & S;
type P = Z;
type Z = { x: string, y: string }
type S = T;
type T = O;
interface O { x: "abc", z: string };
```

results in:
```
    ABC:
      $ref: '#/components/schemas/M'
    M:
      allOf:
        - type: object
          properties:
            x:
              type: string
              enum:
                - abc
            'y':
              type: string
            z:
              type: string
          required:
            - x
            - 'y'
            - z
    O:
      type: object
      properties:
        x:
          type: string
          enum:
            - abc
        z:
          type: string
      required:
        - x
        - z
    P:
      $ref: '#/components/schemas/Z'
    S:
      $ref: '#/components/schemas/T'
    T:
      $ref: '#/components/schemas/O'
    Z:
      type: object
      properties:
        x:
          type: string
        'y':
          type: string
      required:
        - x
        - 'y'
```

Ideally, since M is the narrowed down type, the other elements are not required. 

If you have suggestions on how to do that, that would be great, though it is a little interesting for me to updating what the user may have entered and creating a new object out of it. 

I have pushed the code until the fixes, excluding the code which executes the narrow types above.

Reply:
I'm not sure I understand, for the purposes of `possibleRootTypes`, we should narrow types. However, we don't need to narrow types for OpenAPI output. To clarify I would still expect:
```
  M:
      allOf:
        - $ref: '#/components/schemas/P'
        - $ref: '#/components/schemas/S'
```

Reply:
@mahirk tagging in case you missed ☝️

Reply:
Yup! sorry @lfportal thank you. I updated it so it does not update the output, which made it a lot easier. Now it only combines them for `possibleRootTypes` which allows us to correctly manage the scenarios in https://github.com/airtasker/spot/pull/1133#discussion_r551862324